lib/src/native/realm_core.dart

```diff
@@ -1787,19 +1802,63 @@ final _nativeFinalizer = NativeFinalizer(_realmLib.addresses.realm_release);
 
 abstract class HandleBase<T extends NativeType> implements Finalizable {
   final Pointer<T> _pointer;
+  bool released = false;
+  final bool isUnowned;
 
   @pragma('vm:never-inline')
   void keepAlive() {}
 
-  HandleBase(this._pointer, int size) {
+  HandleBase(this._pointer, int size) : isUnowned = false {
     _nativeFinalizer.attach(this, _pointer.cast(), detach: this, externalSize: size);
-    if (_enableFinalizerTrace) _setupFinalizationTrace(this, _pointer);
+    if (_enableFinalizerTrace) {
+      _setupFinalizationTrace(this, _pointer);
+    }
   }
 
-  HandleBase.unowned(this._pointer);
+  HandleBase.unowned(this._pointer) : isUnowned = true;
 
   @override
   String toString() => "${_pointer.toString()} value=${_pointer.cast<IntPtr>().value}";
+
+  /// @nodoc
+  /// A method that will be invoked just before the handle is released. Allows to cleanup
+  /// any custom data that inheritors are storing.
+  void _releaseCore() {}
+
+  void release() {
+    if (released) {
+      return;
+    }
+
+    _releaseCore();
+
+    if (!isUnowned) {
+      _nativeFinalizer.detach(this);
+      _realmLib.realm_release(_pointer.cast());
+    }
+
+    released = true;
+
+    if (_enableFinalizerTrace) {
+      _tearDownFinalizationTrace(this, _pointer);
+    }
+  }
+}
+
+abstract class RootedHandleBase<T extends NativeType> extends HandleBase<T> {
+  final RealmHandle? _root;
+  int? _id;
+
+  RootedHandleBase(this._root, Pointer<T> pointer, int size) : super(pointer, size) {
+    _id = _root?.addChild(this);
```

Comment:
I'm not sure I see how. The reason for the int is to avoid a situation where the root handle holds a reference to its children. If we keep the current design where the children are kept as weak references, then we need to iterate all children to try and find the one being removed, which seems more inefficient.